aspnetcore/blazor/security/webassembly/standalone-with-identity/qrcodes-for-authenticator-apps.md

```diff
@@ -0,0 +1,853 @@
+                            <p>
+                                <a href="forgot-password">Forgot password</a>
+                            </p>
+                            <p>
+                                <a href="register">Register as a new user</a>
+                            </p>
+                        </div>
+                    </EditForm>
+                </section>
+            </div>
+        </div>
+    </NotAuthorized>
+</AuthorizeView>
+
+@code {
+    private FormResult formResult = new();
+    private bool requiresTwoFactor;
+    private string? recoveryCodesRemainingMessage;
+
+    [SupplyParameterFromForm]
+    private InputModel Input { get; set; } = new();
+
+    [SupplyParameterFromQuery]
+    private string? ReturnUrl { get; set; }
+
+    public async Task LoginUser()
+    {
+        if (requiresTwoFactor)
+        {
+            if (!string.IsNullOrEmpty(Input.TwoFactorCode))
+            {
+                if (Input.TwoFactorCode.Length == 6)
+                {
+                    formResult = await Acct.LoginTwoFactorCodeAsync(
+                        Input.Email, Input.Password, Input.TwoFactorCode);
+                }
+                else
+                {
+                    formResult = await Acct.LoginTwoFactorRecoveryCodeAsync(
+                        Input.Email, Input.Password, Input.TwoFactorCode);
+
+                    if (formResult.Succeeded)
+                    {
+                        var twoFactorResponse = await Acct.TwoFactorRequestAsync(new());
+                        recoveryCodesRemainingMessage =
+                            $"You have {twoFactorResponse.RecoveryCodesLeft} recovery " +
+                            "codes remaining.";
+                    }
+                }
+            }
+        }
+        else
+        {
+            formResult = await Acct.LoginAsync(Input.Email, Input.Password);
+            requiresTwoFactor = formResult.ErrorList.Contains("RequiresTwoFactor");
+            Input.TwoFactorCode = string.Empty;
+            formResult.ErrorList = [];
+        }
+
+        if (formResult.Succeeded && !string.IsNullOrEmpty(ReturnUrl))
+        {
+            Navigation.NavigateTo(ReturnUrl);
+        }
+    }
+
+    private sealed class InputModel
+    {
+        [Required]
+        [EmailAddress]
+        [Display(Name = "Email")]
+        public string Email { get; set; } = string.Empty;
+
+        [Required]
+        [DataType(DataType.Password)]
+        [Display(Name = "Password")]
+        public string Password { get; set; } = string.Empty;
+
+        [Required]
+        [RegularExpression(@"^([0-9]{6})|([A-Z0-9]{5}[-]{1}[A-Z0-9]{5})$", 
+            ErrorMessage = "Must be a six-digit authenticator code (######) or " +
+            "eleven-character alphanumeric recovery code (#####-#####, dash " +
+            "required)")]
+        [Display(Name = "Two-factor code")]
+        public string TwoFactorCode { get; set; } = "123456";
```

Comment:
```suggestion
        [Display(Name = "Two-factor Code or Recovery Code")]
        public string TwoFactorOrRecoveryCode { get; set; } = string.Empty;
```

It's probably best to have the display name match the label and have it start out empty like the other input fields. I think we should also rename the property to make it clearer it can be either thing.

Reply:
I believe that I did this on purpose so that the 2FA field validators don't fire when submitting the initial login request.

I'll try it your way and see if that's the case.